interface/usergroup/usergroup_admin.php

```diff
@@ -147,27 +149,63 @@
+            $olduf = array();
+            while ($tmprow = sqlFetchArray($tmpres)) {
+                $olduf[$tmprow['facility_id'] . '/' . $tmprow['warehouse_id']] = true;
             }
-            array_unshift($sqlBindArray, $_POST["id"]);
-            sqlStatement("delete from users_facility
-            where tablename='users'
-            and table_id= ?
-            and facility_id not in (" . $scheduledFacilityString . ")", $sqlBindArray);
-
+            // Now process the selection of facilities and warehouses.
             foreach ($_POST["schedule_facility"] as $tqvar) {
-                sqlStatement("replace into users_facility set
-                facility_id = ?,
-                tablename='users',
-                table_id = ?", array($tqvar, $_POST["id"]));
+                if (($i = strpos($tqvar, '/')) !== false) {
+                    $facid = substr($tqvar, 0, $i);
+                    $whid = substr($tqvar, $i + 1);
+                    // If there was also a facility-only selection for this warehouse then remove it.
+                    if (isset($olduf["$facid/"])) {
+                        $olduf["$facid/"] = true;
+                    }
+                } else {
+                    $facid = $tqvar;
+                    $whid = '';
+                }
+                if (!isset($olduf["$facid/$whid"])) {
+                    sqlStatement(
+                        "INSERT INTO users_facility SET tablename = ?, table_id = ?, " .
+                        "facility_id = ?, warehouse_id = ?",
+                        array('users', $_POST["id"], $facid, $whid)
+                    );
+                }
+                $olduf["$facid/$whid"] = false;
+                if ($facid == $deffacid) {
+                    $deffacid = 0;
+                }
+            }
+            // Now delete whatever is left over for this user.
+            foreach ($olduf as $key => $value) {
+                if ($value && ($i = strpos($key, '/')) !== false) {
+                    $facid = substr($key, 0, $i);
+                    $whid = substr($key, $i + 1);
+                    sqlStatement(
+                        "DELETE FROM users_facility WHERE " .
+                        // The following screws up by matching all warehouse_id values when it's
+                        // an empty string. This needs debugging in sql.inc.
+                        /**********************************************
+                        "tablename = ? AND table_id = ? AND facility_id = ? AND warehouse_id = ?",
+                        array('users', $_POST["id"], $facid, $whid));
+                        **********************************************/
+                        "tablename = 'users' AND table_id = " . add_escape_custom($_POST["id"]) .
+                        " AND facility_id = '" . add_escape_custom($facid) . "'" .
+                        " AND warehouse_id = '" . add_escape_custom($whid) . "'"
```

Comment:
better to use binding in this query

Reply:
See my comment above these 3 lines. I'll change it back in case whatever that bug was is fixed.

Reply:
weird stuff. if you need to add it back, just will need quotes in query around that add_escape_custom($_POST["id"] call